### PR TITLE
Add option to render manifest from previous build result

### PR DESCRIPTION
**Pull request by AndiDog**
Relates to #1187, #922, #1937

Should merge before : n/a

Should merge after : n/a

**Description**

Add option to render manifest from previous build result.

This is helpful e.g. in scenarios where build should be run just once, but manifest generation multiple times (for instance, once for dev/staging/prod).

**User facing changes**

Backwards-compatible: `skaffold render` first builds, unless the new parameter `--build-artifacts` is given – then it reads the images from the existing build output file.

**Next PRs**

Only slightly related: potentially, an option `--no-build` would make sense to render the manifest as specified in `deploy:` section, but without building and replacing image tags. That could be used in CI to fail early if the manifest / Helm chart isn't valid – _before_ running the build which could take several minutes and incur network/compute resources.

**Hints for reviewer**

Didn't see a good place to add tests, since my implementation is at `cmd` level. Please advise and I'll happily add a test.

**Submitter Checklist**

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [unit tests](../DEVELOPMENT.md#creating-a-pr)
- [ ] Mentions any output changes.
- [ ] Adds documentation as needed: user docs, YAML reference, CLI reference.
- [ ] Adds integration tests if needed.

**Reviewer Notes**

- [ ] The code flow looks good. 
- [ ] Unit test added.
- [ ] User facing changes look good.

**Release Notes**

```
`skaffold render` now supports a parameter `--build-artifacts` to take image tags from a previous file (as created by `skaffold build --file-output=...`). Without this parameter, the default behavior is to build the images first.
```

### CODE REVIEW

**Comment by AndiDog** [comment]
Changed parameter name to `-a/--build-artifacts` for consistency with `skaffold deploy`.

**Comment by balopat** [comment]
Thanks @AndiDog, can you please run `./hack/generate-man.sh` and commit the results?

**Comment by AndiDog** [comment]
@balopat do you have a hint where a test could be placed? My implementation is at `cmd` level, so `integration/render_test.go` doesn't really fit since the `builds: []build.Artifact{ ... }` is already filled there, instead of taken from a file. Ideally, there would be a test that runs the `skaffold` executable directly instead of running a Go test function. I could add a test for `flags.BuildOutputFileFlag :: func BuildArtifacts` which is what the command implementation uses?!

**Comment by balopat** [comment]
> @balopat do you have a hint where a test could be placed? My implementation is at `cmd` level, so `integration/render_test.go` doesn't really fit since the `builds: []build.Artifact{ ... }` is already filled there, instead of taken from a file. Ideally, there would be a test that runs the `skaffold` executable directly instead of running a Go test function. I could add a test for `flags.BuildOutputFileFlag :: func BuildArtifacts` which is what the command implementation uses?!

I think adding an integration test is a good incremental step. You can create a new function `TestKubectlRenderFromBuildOutput` in integration/render_test.go that is similar to TestKubectlRender, just takes a build output file as well as an input.

**Comment by AndiDog** [comment]
I modified the command code to make it testable – please have a look. This also makes testing other deployers much easier because the test works almost exactly like running `skaffold render` directly.

Plus new commit "Ignore local executables for integration test (architecture/OS might mismatch)". Makes the kind-based integration test work on macOS for me.

Please have a look if this way of testing looks good.

File: Makefile [hunk 1]
```diff
@@ -171,7 +171,7 @@ build_deps:
 
 .PHONY: skaffold-builder
 skaffold-builder:
-	time docker build \
+	DOCKER_BUILDKIT=1 time docker build \
```
Comment (balopat) [inline Makefile]: hmm...can we separate this change out? we'll need to test (unfortunately manually) the impact of this on our release scripts, I'm not confident this won't break things on GCB.

**Comment by balopat** [comment]
> I modified the command code to make it testable – please have a look. This also makes testing other deployers much easier because the test works almost exactly like running `skaffold render` directly.

Thanks for doing this work! 
However -  I don't see the benefit yet: integration tests are supposed to call the built skaffold binary with flags. The three things you pass in [here](https://github.com/GoogleContainerTools/skaffold/pull/3567/files#diff-d3f21d5008ea0ca75c6dfac3afbce9a4R245-R247) and the buildoutput flag should all be available as flags to the skaffold command.

**Comment by balopat** [comment]
> Plus new commit "Ignore local executables for integration test (architecture/OS might mismatch)". Makes the kind-based integration test work on macOS for me.

Thanks - this is cool. I don't know when I ran kind integration tests on my mac the last time :) probably that's why you ran into this!

**Comment by balopat** [comment]
Actually we need to remove the BUILDKIT env var, our kokoro test is failing due to this here: 

```
++ gcloud auth configure-docker -q
Docker configuration file updated.
++ '[' -e /tmpfs/src/gfile/bazel-cache-gcs ']'
+ pushd /tmpfs/src/github/skaffold
+ GCP_ONLY=true
+ make integration-in-docker
DOCKER_BUILDKIT=1 time docker build \
		-f deploy/skaffold/Dockerfile \
		-t gcr.io/k8s-skaffold/skaffold-builder \
		.
buildkit not supported by daemon
Command exited with non-zero status 1
```

**Comment by AndiDog** [comment]
Working on the changes.

I noted that `skaffold render` connects to the Kubernetes cluster in the `renderManifests` function. That shouldn't be the case, or it's totally unusable for CI use cases. `skaffold render` should only render, i.e. do text replacement in an existing manifest, nothing else. That should be offline. Have to fix that as well before I can push the changes for your review.

**Comment by AndiDog** [comment]
Hmm, `kubectl create` seems to always connect to a cluster. We're kind of screwed here. What about we replace that by not calling `kubectl` and simply making a list of manifests in the `readManifests` function (at least for the offline case, i.e. `skaffold render`)?

**Comment by nkubala** [comment]
@AndiDog i'd love to help you get this merged soon. i'm trying to catch up on the conversation, but can you explain the biggest blocker you're having right now? also, could you please rebase?

**Comment by AndiDog** [comment]
Don't worry. I'm working on it **right now**. Rebased and found a good way to add tests. I've also resolved the problem that a cluster is needed to render manifests – my next update will contain options `--offline` (don't talk to cluster, use manifests as-are) and `--no-labels` (avoid adding Skaffold-specific labels) which are both required for GitOps use cases where the manifest is created at a separate stage and Skaffold isn't the deployer. Give me a few more hours or days to get it all working and cleaned 😉 I will get back with questions if there are more blockers on the road.

**Comment by AndiDog** [comment]
```
func dumpToFileOrWriter(renderedManifests string, filepath string, manifestOut io.Writer) error {
	if filepath == "" {
		_, err := fmt.Fprintln(manifestOut, renderedManifests)
		return err
	}

	f, err := os.OpenFile(filepath, os.O_RDWR|os.O_CREATE, 0666)
	if err != nil {
		return errors.Wrap(err, "opening file for writing manifests")
	}
	defer f.Close()
	_, err = f.WriteString(renderedManifests + "\n")
	return err
}
```

This doesn't truncate the output file, i.e. leaves existing file content at the end. Not typically an issue unless you work for days on a PR and are already confused enough 😂 Will fix this as well.

**Comment by nkubala** [comment]
@AndiDog awesome! just ping me on here when you're ready for me to review :)

**Comment by AndiDog** [comment]
@nkubala @balopat Lints and tests fixed. Finished from my side. Please go ahead and review.

**Comment by AndiDog** [comment]
Rebased to solve a simple merge conflict. Should we continue here? Any blockers or objections?

**Comment by alnr** [comment]
Bump. Would really like to see this feature. 👍

**Comment by AndiDog** [comment]
@nkubala @balopat can you please review? I rebased it once again, and tests succeed.

File: pkg/skaffold/deploy/helm.go [hunk 2]
```diff
@@ -62,9 +63,7 @@ type HelmDeployer struct {
 	*latest.HelmDeploy
 
 	kubeContext string
-	kubeConfig  string
-	namespace   string
-	forceDeploy bool
+	opts        config.SkaffoldOptions
```
Comment (nkubala) [inline pkg/skaffold/deploy/helm.go:66]: we try where we can to avoid passing around resources that aren't necessary. is there a certain field you need from the `SkaffoldOptions` that you could grab out and embed in the `HelmDeployer` rather than just putting the whole opts in there?

File: pkg/skaffold/deploy/labels.go [hunk 3]
```diff
@@ -48,7 +49,11 @@ type Labeller interface {
 }
 
 // merge merges the labels from multiple sources.
-func merge(deployer Labeller, sources ...Labeller) map[string]string {
+func merge(opts config.SkaffoldOptions, deployer Labeller, sources ...Labeller) map[string]string {
+	if opts.NoLabels {
```
Comment (nkubala) [inline pkg/skaffold/deploy/labels.go:53]: we don't need to pass the entire `SkaffoldOptions` here. just passing the `NoLabels` bool should be enough

File: pkg/skaffold/deploy/kustomize.go [hunk 4]
```diff
@@ -197,15 +200,15 @@ func (k *KustomizeDeployer) Dependencies() ([]string, error) {
...
+func (k *KustomizeDeployer) Render(ctx context.Context, out io.Writer, builds []build.Artifact, labellers []Labeller, offline bool, filepath string) error {
 	manifests, err := k.renderManifests(ctx, out, builds, labellers)
 	if err != nil {
 		return err
 	}
 
 	manifestOut := out
 	if filepath != "" {
-		f, err := os.OpenFile(filepath, os.O_RDWR|os.O_CREATE, 0666)
+		f, err := os.OpenFile(filepath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0666)
```
Comment (nkubala) [inline pkg/skaffold/deploy/kustomize.go:211]: any particular reason we needed to do this? I'm not opposed to it, just wondering.

Comment (AndiDog) [reply pkg/skaffold/deploy/kustomize.go:211]: If the file already exists with content `ABC`, and here you write content `D`, you'd get a file containing `DBC` without this fix.

File: pkg/skaffold/deploy/helm.go (hunk 2 above)
Comment (AndiDog) [reply pkg/skaffold/deploy/helm.go:66]: Rolled back those changes

**Comment by AndiDog** [comment]
@nkubala can you review?

**Comment by alnr** [comment]
Bump 👍

File: cmd/skaffold/app/cmd/render.go [hunk 5]
```diff
@@ -40,8 +44,12 @@ func NewCmdRender() *cobra.Command {
 		WithDescription("[alpha] Perform all image builds, and output rendered Kubernetes manifests").
 		WithCommonFlags().
 		WithFlags(func(f *pflag.FlagSet) {
-			f.BoolVar(&showBuild, "loud", false, "Show the build logs and output")
+			f.VarP(&fromBuildOutputFile, "build-artifacts", "a", "Don't build, but use file created with `skaffold build --file-output`")
+			f.BoolVar(&offline, "offline", false, `Do not connect to Kubernetes server API for manifest creation and validation.
+This is helpful when no Kubernetes cluster is available (e.g. GitOps model).
+No metadata.namespace attribute is injected in this case - the manifest content does not get changed.`)
 			f.StringVar(&renderOutputPath, "output", "", "file to write rendered manifests to")
+			f.BoolVar(&showBuild, "loud", false, "Show the build logs and output")
```
Comment (tstromberg) [inline cmd/skaffold/app/cmd/render.go:52]: This flag seems ambiguous and non-obvious - is it feasible to move into a separate PR for its own discussion? 

I feel like this should be called `--show-build-logs` or `--exit-after-logs`?

File: cmd/skaffold/app/cmd/flags.go [hunk 6]
```diff
@@ -277,6 +277,14 @@ var FlagRegistry = []Flag{
 		FlagAddMethod: "BoolVar",
 		DefinedOn:     []string{"dev", "run", "debug", "deploy", "render", "build", "delete", "diagnose"},
 	},
+	{
+		Name:          "no-labels",
```
Comment (tstromberg) [inline cmd/skaffold/app/cmd/flags.go:281]: Please do not add negative boolean flags. They cause a lot of confusion, when someone says:

`--no-labels=false`

Instead, should this be `--add-skaffold-labels=false`?

File: pkg/skaffold/deploy/kubectl.go [hunk 7]
```diff
@@ -170,24 +174,44 @@ func (k *KubectlDeployer) manifestFiles(manifests []string) ([]string, error) {
...
 			manifests = append(manifests, manifest)
+			hasURLManifest = true
 		}
 	}
 
 	if len(manifests) == 0 {
 		return deploy.ManifestList{}, nil
 	}
 
+	// In case no URLs are provided, we can stay offline - no need to run "kubectl create" which would try to connect to a cluster
```
Comment (tstromberg) [inline pkg/skaffold/deploy/kubectl.go:197]: Please handle the easy flow early so that the exception here can be de-idented as much as possible.

`if !offline {
  return k.kubectl.ReadManifests(ctx, manifests)
}

...
do other things
`

File: pkg/skaffold/config/options.go [hunk 8]
```diff
@@ -51,22 +51,28 @@ type SkaffoldOptions struct {
...
-	Profiles              []string
-	InsecureRegistries    []string
-	Command               string
-	RPCPort               int
-	RPCHTTPPort           int
+
+	// Avoids setting Skaffold-specific labels including runID, deployer labels, etc.
+	// `CustomLabels` are still applied. Must only be used in commands which don't deploy
+	// (e.g. `skaffold render`) since the runID label isn't available.
+	NoLabels bool
```
Comment (tstromberg) [inline pkg/skaffold/config/options.go:58]: As I mentioned elsewhere, try to keep bools in the positive. This should be "Labels" with a false, for instance.

`Labels` seems overly ambiguous here, given the huge amount of description required.

File: cmd/skaffold/app/cmd/render.go [hunk 9]
```diff
@@ -25,13 +25,17 @@ import (
...
+	"github.com/GoogleContainerTools/skaffold/cmd/skaffold/app/flags"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/build"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/runner"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
 )
 
 var (
-	showBuild        bool
-	renderOutputPath string
+	fromBuildOutputFile flags.BuildOutputFileFlag
```
Comment (balopat) [inline cmd/skaffold/app/cmd/render.go:35]: We should limit this PR to this flag. 
Please split `offline`, `renderOutputPath`, `showBuild` should go to individual PRs - it is going to be much faster to review.

File: cmd/skaffold/app/cmd/render.go (hunk 5 above)
Comment (AndiDog) [reply cmd/skaffold/app/cmd/render.go:52]: The `--loud` flag was there before. I only reordered the arguments alphabetically (not really, though 🤦‍♂️, sorry). Out of scope for this review. I'll change the order back so the diff looks sane.

File: cmd/skaffold/app/cmd/flags.go (hunk 6 above)
Comment (AndiDog) [reply cmd/skaffold/app/cmd/flags.go:281]: Will change to `--add-skaffold-labels=true` as default

File: cmd/skaffold/app/cmd/render.go (hunk 9 above)
Comment (AndiDog) [reply cmd/skaffold/app/cmd/render.go:35]: Sorry, I have confused you by ordering those lines alphabetically. The only _new_ flags are `fromBuildOutputFile`/`offline` which I believe should be reviewed together to make sense for the GitOps model. I'm reverting the order so the diff looks sane, and will follow @tstromberg's advice on naming.

File: pkg/skaffold/config/options.go (hunk 8 above)
Comment (AndiDog) [reply pkg/skaffold/config/options.go]: Changed to `AddSkaffoldLabels bool`

**Comment by AndiDog** [comment]
@balopat @nkubala @tstromberg All review requests solved, and rebased onto the new Helm render feature.

File: cmd/skaffold/app/cmd/render.go [hunk 10]
```diff
@@ -41,6 +45,8 @@ func NewCmdRender() *cobra.Command {
 		WithCommonFlags().
 		WithFlags(func(f *pflag.FlagSet) {
 			f.BoolVar(&showBuild, "loud", false, "Show the build logs and output")
+			f.VarP(&fromBuildOutputFile, "build-artifacts", "a", "Don't build, but use file created with `skaffold build --file-output`")
+			f.BoolVar(&offline, "offline", false, `Do not connect to Kubernetes server API for manifest creation and validation. This is helpful when no Kubernetes cluster is available (e.g. GitOps model). No metadata.namespace attribute is injected in this case - the manifest content does not get changed.`)
```
Comment (nkubala) [inline cmd/skaffold/app/cmd/render.go:49]: nit: 

```suggestion
			f.BoolVar(&offline, "offline", false, `Do not connect to Kubernetes API server for manifest creation and validation. This is helpful when no Kubernetes cluster is available (e.g. GitOps model). No metadata.namespace attribute is injected in this case - the manifest content does not get changed.`)
```

File: cmd/skaffold/app/cmd/render.go [hunk 11]
```diff
@@ -41,6 +45,8 @@ func NewCmdRender() *cobra.Command {
 		WithCommonFlags().
 		WithFlags(func(f *pflag.FlagSet) {
 			f.BoolVar(&showBuild, "loud", false, "Show the build logs and output")
+			f.VarP(&fromBuildOutputFile, "build-artifacts", "a", "Don't build, but use file created with `skaffold build --file-output`")
```
Comment (nkubala) [inline cmd/skaffold/app/cmd/render.go:48]: this feels weird to use: I feel like with this flag I'd be providing the list of artifacts on the command line, not a file. can we change this to something like `--build-file`?

Comment (nkubala) [reply cmd/skaffold/app/cmd/render.go:48]: also, for the description I'd say something like `File containing build result from a previous 'skaffold build'`

File: cmd/skaffold/app/cmd/render.go [hunk 12]
```diff
@@ -25,13 +25,17 @@ import (
...
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/runner"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
 )
 
 var (
-	showBuild        bool
-	renderOutputPath string
+	showBuild           bool
+	renderOutputPath    string
+	fromBuildOutputFile flags.BuildOutputFileFlag
```
Comment (nkubala) [inline cmd/skaffold/app/cmd/render.go:37]:
```suggestion
	buildOutputFile flags.BuildOutputFileFlag
```

File: cmd/skaffold/app/cmd/render.go (hunk 11 above)
Comment (AndiDog) [reply cmd/skaffold/app/cmd/render.go:48]: Also not happy with the name but this parameter has existed for a long time in `skaffold deploy`, so I kept the name consistent.

I propose we either keep as-is, or change `skaffold deploy` to support both old and new name roughly like this (untested):

```go
func NewCmdDeploy() *cobra.Command {
	return NewCmd("deploy").
		// [...]
		WithFlags(func(f *pflag.FlagSet) {
			// [...]
			f.VarP(&buildOutputFile, "build-artifacts", "", "Deploy with images in build output file created with `skaffold build --file-output`")
			f.VarP(&buildOutputFile, "build-file", "a", "Deploy with images in build output file created with `skaffold build --file-output`")
			}).
		// [...]
}
```

File: cmd/skaffold/app/cmd/render.go (hunk 12 above)
Comment (AndiDog) [reply cmd/skaffold/app/cmd/render.go:37]: `var` blocks are global in Go, and there's already such a variable for the deploy command... I'll make both variable names unique `{deploy,render}FromBuildOutputFile` unless you have a better idea.

File: cmd/skaffold/app/cmd/render.go (hunk 11 above)
Comment (nkubala) [reply cmd/skaffold/app/cmd/render.go]: eh, yeah ok I didn't realize this isn't new. I guess let's just leave it the way it's been for now, we can change it later.

File: cmd/skaffold/app/cmd/render.go (hunk 12 above)
Comment (nkubala) [reply cmd/skaffold/app/cmd/render.go]: `deployBuildOutputFile` and `renderBuildOutputFile` seem fine

**Comment by nkubala** [comment]
@AndiDog rename those two vars and rebase to get CI green and I'll merge this!

**Comment by AndiDog** [comment]
@nkubala ready!
